File: other/w4a8_linear.py
# -*- coding: utf-8 -*-
"""W4A8 量化 Linear 算子(AWQ 风格权重布局,纯 PyTorch 实现)。

## AWQ 4-bit 权重布局

与 GPTQ 沿 K 方向 pack 不同,AWQ 为了让 dequant kernel 一次性取出 low/high
两组 nibble,采用沿 **N 方向** pack 并带有 **interleave 重排**的约定:

- ``qweight``: int32, shape ``[K, N // 8]``。
  沿 **N 方向** 每 8 个 4-bit 权重打包进一个 int32;**nibble 的 pack 顺序为
  interleaved**:``[0, 2, 4, 6, 1, 3, 5, 7]``。即:

  * 第 0 个 nibble(bits [0, 4)) 放 N 维第 0 列
  * 第 1 个 nibble(bits [4, 8)) 放 N 维第 2 列
  * 第 2 个 nibble(bits [8, 12))放 N 维第 4 列
  * 第 3 个 nibble(bits [12, 16))放 N 维第 6 列
  * 第 4 个 nibble(bits [16, 20))放 N 维第 1 列
  * 第 5 个 nibble(bits [20, 24))放 N 维第 3 列
  * 第 6 个 nibble(bits [24, 28))放 N 维第 5 列
  * 第 7 个 nibble(bits [28, 32))放 N 维第 7 列

- ``qzeros``:  int32, shape ``[K // group_size, N // 8]``。
  沿 **N 方向** 每 8 个 4-bit 零点打包进一个 int32,interleave 顺序与 ``qweight``
  相同。
- ``scales``:  float16, shape ``[K // group_size, N]``,与 ``qweight`` 的 N 维
  **不做 interleave**(scales 本身就是按列排布的)。

反量化语义:``w_fp[k, j] = (w_q[k, j] - w_z[k // g, j]) * scales[k // g, j]``。

## 计算流程(per-token int8 激活 + per-group int8 权重 + int8 GEMM)

1. 对输入 ``x : [M, K] bf16`` 做 per-token 动态对称 int8 量化,得到
   ``x_q : [M, K] int8`` 与 ``x_scale : [M] float32``。
2. 将 ``qweight`` / ``qzeros`` 解包并反 interleave,得到
   ``w_q : [K, N] int8`` 与 ``w_z : [K//g, N] int8``。
3. 按 K 维每个 group 独立做 int8 GEMM
   ``acc_g = x_q[:, k0:k1] @ (w_q[k0:k1] - w_z[g]) (int32)``,随后以
   ``x_scale * w_scale[g]`` 反量化并累加至 fp32。
4. 转回 bfloat16。

本实现不依赖 vLLM / C++ 扩展,可作为 AWQ 等价性对比的参考路径。
"""
from __future__ import annotations

import torch

__all__ = [
    "AWQ_REVERSE_ORDER",
    "unpack_awq_qweight",
    "unpack_awq_qzeros",
    "w4a8_linear",
]


# AWQ 原始 interleave 顺序:第 i 个 nibble 对应 N 维的 AWQ_ORDER[i] 列
# 权威来源:
#   https://github.com/casper-hansen/AutoAWQ/blob/main/awq/utils/packing_utils.py
#   https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/awq.py
AWQ_ORDER: tuple[int, ...] = (0, 2, 4, 6, 1, 3, 5, 7)

# 逆序:给定输出 N 维第 j 列,应当取第 AWQ_REVERSE_ORDER[j] 个 nibble
# 由 AWQ_ORDER 反推:[0, 4, 1, 5, 2, 6, 3, 7]
AWQ_REVERSE_ORDER: tuple[int, ...] = tuple(
    AWQ_ORDER.index(j) for j in range(8)
)

_BIT_SHIFTS = torch.arange(0, 32, 4, dtype=torch.int32)


# ── 形状校验 ──

def _check_shapes(
    x: torch.Tensor,
    qweight: torch.Tensor,
    qzeros: torch.Tensor,
    scales: torch.Tensor,
) -> tuple[int, int, int, int]:
    """校验输入张量形状并返回 ``(M, K, N, group_size)``。"""
    if x.dtype != torch.bfloat16:
        raise RuntimeError(f"x 必须为 bfloat16,当前 dtype={x.dtype}")
    if qweight.dtype != torch.int32:
        raise RuntimeError(f"qweight 必须为 int32,当前 dtype={qweight.dtype}")
    if qzeros.dtype != torch.int32:
        raise RuntimeError(f"qzeros 必须为 int32,当前 dtype={qzeros.dtype}")
    if scales.dtype != torch.float16:
        raise RuntimeError(f"scales 必须为 float16,当前 dtype={scales.dtype}")

    if qweight.dim() != 2 or qzeros.dim() != 2 or scales.dim() != 2:
        raise RuntimeError("qweight / qzeros / scales 必须均为 2D")

    k, n_over_8 = qweight.shape
    n = n_over_8 * 8

    groups_z, n_over_8_z = qzeros.shape
    groups_s, n_s = scales.shape

    if n_over_8_z != n_over_8:
        raise RuntimeError(
            f"qzeros N 维应与 qweight 一致: {n_over_8_z} vs {n_over_8}"
        )
    if groups_z != groups_s:
        raise RuntimeError(
            f"qzeros 和 scales 的 group 维不一致: {groups_z} vs {groups_s}"
        )
    if n_s != n:
        raise RuntimeError(f"scales 的 N 维应为 {n},当前为 {n_s}")
    if k % groups_z != 0:
        raise RuntimeError(f"K={k} 不能被 group 数 {groups_z} 整除")

    group_size = k // groups_z
    if x.shape[-1] != k:
        raise RuntimeError(f"x 最后一维应为 K={k},当前为 {x.shape[-1]}")

    m = x.numel() // k
    return m, k, n, group_size


# ── AWQ 解包 ──

def _unpack_int4_along_n_awq(packed: torch.Tensor) -> torch.Tensor:
    """沿 N 方向解包 AWQ int32 → int8(含 interleave 反重排)。

    输入 ``packed : [..., N // 8] int32``,输出 ``[..., N] int8``,值域 ``[0, 15]``。
    """
    if packed.dtype != torch.int32:
        raise RuntimeError(f"packed 必须为 int32,当前 dtype={packed.dtype}")

    lead = packed.shape[:-1]
    n_over_8 = packed.shape[-1]

    shifts = _BIT_SHIFTS.to(packed.device)                        # [8]
    # 逐 nibble 提取 → [..., N//8, 8](第 8 维按 pack 内顺序,即 AWQ_ORDER)
    unpacked = (packed.unsqueeze(-1) >> shifts.view(*([1] * len(lead)), 1, 8)) & 0xF

    # 反 interleave:按 AWQ_REVERSE_ORDER 取 nibble,使得输出沿 N 维是自然顺序
    reverse_idx = torch.tensor(
        AWQ_REVERSE_ORDER, dtype=torch.long, device=packed.device,
    )
    unpacked = unpacked.index_select(-1, reverse_idx)             # [..., N//8, 8]
    return unpacked.reshape(*lead, n_over_8 * 8).to(torch.int8)


def unpack_awq_qweight(qweight: torch.Tensor) -> torch.Tensor:
    """解包 AWQ 的 ``qweight [K, N // 8] int32`` 为 ``[K, N] int8``。"""
    if qweight.dim() != 2:
        raise RuntimeError(f"qweight 必须为 2D,当前 {qweight.dim()}D")
    return _unpack_int4_along_n_awq(qweight)


def unpack_awq_qzeros(qzeros: torch.Tensor) -> torch.Tensor:
    """解包 AWQ 的 ``qzeros [G, N // 8] int32`` 为 ``[G, N] int8``。"""
    if qzeros.dim() != 2:
        raise RuntimeError(f"qzeros 必须为 2D,当前 {qzeros.dim()}D")
    return _unpack_int4_along_n_awq(qzeros)


# ── 激活量化 ──

def _quantize_activation_per_token(x: torch.Tensor,) -> tuple[torch.Tensor, torch.Tensor]:
    """对 ``x [M, K]`` 做 per-token 动态对称 int8 量化。

    :returns: ``(x_q [M, K] int8, x_scale [M] float32)``
    """
    x_fp32 = x.to(torch.float32)
    max_abs = x_fp32.abs().amax(dim=-1)                           # [M]
    # max_abs 为 0 的行 scale 置 1.0 以避免除零(此时 x_q 为全 0)
    scale = torch.where(
        max_abs > 0,
        max_abs / 127.0,
        torch.ones_like(max_abs),
    )
    x_q = torch.round(x_fp32 / scale.unsqueeze(-1)).clamp_(-128, 127).to(torch.int8)
    return x_q, scale


# ── 主算子 ──

def w4a8_linear(
    x: torch.Tensor,
    qweight: torch.Tensor,
    qzeros: torch.Tensor,
    scales: torch.Tensor,
    bias: torch.Tensor | None = None,
) -> torch.Tensor:
    """W4A8 量化 Linear(AWQ 布局):per-token int8 GEMM 实现。

    在数值上等价于:

    .. code-block:: python

        w_q = unpack_awq_qweight(qweight).float()                 # [K, N]
        w_z = unpack_awq_qzeros(qzeros).float()                   # [G, N]
        w_z_full = w_z.repeat_interleave(group_size, dim=0)       # [K, N]
        s_full = scales.float().repeat_interleave(group_size, 0)  # [K, N]
        w_fp = (w_q - w_z_full) * s_full
        out = (x.float() @ w_fp).to(bfloat16)

    但内部以 **per-token int8 激活 + per-group int8 权重 + int8 GEMM** 路径计算,
    贴近生产侧 W4A8 算子的数值特征。

    :param x:       bfloat16 张量,形状 ``[M, K]``。
    :param qweight: int32 张量,形状 ``[K, N // 8]``,AWQ 沿 N 方向 interleaved pack。
    :param qzeros:  int32 张量,形状 ``[K // group_size, N // 8]``,AWQ interleaved pack。
    :param scales:  float16 张量,形状 ``[K // group_size, N]``。
    :param bias:    可选 bias,形状 ``[N]``,输出 dtype 为 bfloat16。
    :returns:       bfloat16 张量,形状 ``[*, N]``。
    """
    _m, k, n, group_size = _check_shapes(x, qweight, qzeros, scales)

    # 空输入短路
    if x.numel() == 0:
        return torch.empty(
            (*x.shape[:-1], n), dtype=torch.bfloat16, device=x.device,
        )

    orig_shape = x.shape
    x_2d = x.contiguous().reshape(-1, k)                          # [M, K] bf16

    # 1) 激活 per-token int8 量化
    x_q, x_scale = _quantize_activation_per_token(x_2d)           # int8, fp32

    # 2) AWQ 权重 / 零点解包(一次性到 int8 自然排布的 [K, N] / [G, N])
    w_q = unpack_awq_qweight(qweight.contiguous())                # [K, N] int8
    w_z = unpack_awq_qzeros(qzeros.contiguous())                  # [G, N] int8
    w_scale = scales.contiguous().to(torch.float32)               # [G, N] fp32

    # 3) per-group int8 GEMM,fp32 累加
    groups_k = k // group_size
    out_fp32 = torch.zeros(
        (x_q.shape[0], n), dtype=torch.float32, device=x.device,
    )
    for g in range(groups_k):
        k0 = g * group_size
        k1 = k0 + group_size

        # 本 group 的中心化权重:w_q - w_z ∈ [-15, 15],可容纳于 int8
        w_block_i32 = (
            w_q[k0:k1].to(torch.int32)
            - w_z[g].to(torch.int32).unsqueeze(0)
        )                                                         # [gs, N]

        # torch.matmul 不支持 int8;以 int32 提升保证位宽不溢出(int8*int8+ 累加)
        # 替换为int8 gemm
        acc = torch.matmul(
            x_q[:, k0:k1].to(torch.int32),
            w_block_i32,
        )                                                         # [M, N] int32

        out_fp32.add_(acc.to(torch.float32) * w_scale[g].unsqueeze(0))

    # 激活 scale 在最外层统一乘一次,避免 per-group 重复广播
    out_fp32.mul_(x_scale.unsqueeze(-1))

    if bias is not None:
        if bias.shape != (n,):
            raise RuntimeError(
                f"bias 形状应为 ({n},),当前为 {tuple(bias.shape)}"
            )
        out_fp32.add_(bias.to(torch.float32))

    return out_fp32.to(torch.bfloat16).reshape(*orig_shape[:-1], n)

File: other/w4a8_linear.cpp
#include <cstdint>
#include <vector>
#include <stdexcept>
#include <cmath>
#include <algorithm>
#include <string>
#include <arm_neon.h>   // float16_t
#include <arm_bf16.h>   // __bf16 / bfloat16_t (ARM BF16 扩展)

namespace w4a8 {
    constexpr int AWQ_ORDER[8] = {0, 2, 4, 6, 1, 3, 5, 7};
    constexpr int AWQ_REVERSE_ORDER[8] = {0, 4, 1, 5, 2, 6, 3, 7};
    constexpr int BIT_SHIFTS[8] = {0, 4, 8, 12, 16, 20, 24, 28};

    struct LinearParams {
        int M;          // batch size（token 数）
        int K;          // 输入特征维度
        int N;          // 输出特征维度
        int group_size; // 量化 group 大小
        int groups;     // group 数量 = K / group_size
    };
    inline LinearParams check_shapes(int x_rows, int x_cols,           // x:       [M, K]
                                    int qw_rows, int qw_cols,         // qweight: [K, N/8]
                                    int qz_rows, int qz_cols,         // qzeros:  [groups, N/8]
                                    int sc_rows, int sc_cols) {          // scales:  [groups, N]
        int K = x_cols;
        int M = x_rows;
        
        int n_over_8 = qw_cols;
        int N = n_over_8 * 8;
        
        if (qw_rows != K) {
            throw std::runtime_error("qweight rows must equal K");
        }
        if (qz_cols != n_over_8) {
            throw std::runtime_error(
                "qzeros N dim mismatch: " + std::to_string(qz_cols) +
                " vs " + std::to_string(n_over_8));
        }
        if (qz_rows != sc_rows) {
            throw std::runtime_error(
                "qzeros and scales group dim mismatch: " +
                std::to_string(qz_rows) + " vs " + std::to_string(sc_rows));
        }
        if (sc_cols != N) {
            throw std::runtime_error(
                "scales N dim mismatch: " + std::to_string(sc_cols) +
                " vs " + std::to_string(N));
        }
        if (K % qz_rows != 0) {
            throw std::runtime_error(
                "K=" + std::to_string(K) +
                " not divisible by groups=" + std::to_string(qz_rows));
        }

        int group_size = K / qz_rows;
        int groups = qz_rows;
        return {M, K, N, group_size, groups};
    }
    std::pair<std::vector<int8_t>, std::vector<float>> quantize_activation_per_token(const __bf16* x, int M, int K) {
        std::vector<int8_t> x_q(M * K);
        std::vector<float> x_scale(M);

        std::vector<float> max_abs(M, 0.0f);
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < K; j++) {
                float v = static_cast<float>(x[i * K + j]);   // bf16 -> fp32
                max_abs[i] = std::max(max_abs[i], std::abs(v));
            }
            if(max_abs[i] == 0.0f) {
                max_abs[i] = 1.0f;
            }
            x_scale[i] = max_abs[i] / 127.0f;
            float inv_scale = 1.0f / x_scale[i];
            for (int j = 0; j < K; j++) {
                float v = static_cast<float>(x[i * K + j]) * inv_scale;
                v = std::round(v);
                v = std::max(-128.0f, std::min(127.0f, v));
                x_q[i * K + j] = static_cast<int8_t>(v);
            }
        }
        return {x_q, x_scale};
    }

    std::vector<int8_t> unpack_int4_along_n_awq(const int32_t* packed, int rows, int n_over_8) {
        int N = n_over_8 * 8;
        std::vector<int8_t> result(rows * N);
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < n_over_8; j++) {
                int32_t val = packed[i * n_over_8 + j]; 
                int8_t nibbles[8];
                //取出8个4bit
                for(int k = 0; k < 8; k++) {
                    nibbles[k] = (val >> (k * 4)) & 0xF;   //(k * 4) = BIT_SHIFTS[k]
                }
                //按照AWQ_REVERSE_ORDER的顺序写入result
                for(int k = 0; k < 8; k++) {
                    result[i * N + j * 8 + k] = nibbles[AWQ_REVERSE_ORDER[k]];
                }
            }
        }
        return result;
    }
    std::vector<int8_t> unpack_awq_qweight(const int32_t* packed, int K, int n_over_8) {
        return unpack_int4_along_n_awq(packed, K, n_over_8);
    }
    std::vector<int8_t> unpack_awq_qzeros(const int32_t* packed, int K, int n_over_8) {
        return unpack_int4_along_n_awq(packed, K, n_over_8);
    }

    void gemm_int32(const int32_t* a, const int32_t* b, int32_t* c, int m, int k, int n) {
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                int32_t sum = 0;
                for(int kk = 0; kk < k; kk++) {
                    sum += a[i * k + kk] * b[kk * n + j];
                }
                c[i * n + j] = sum;
            }
        }
    }


    std::vector<float> w4a8_linear(const __bf16* x, int M, int K,
                                const int32_t* qweight, int N, 
                                const int32_t* qzeros, int group_size, 
                                const float16_t* scales, const __bf16* bias) {

        int n_over_8 = N / 8;
        int groups = K / group_size;
        auto params = check_shapes(M, K, K, n_over_8, groups, n_over_8, groups, N);
        if(M == 0 || K == 0) {
            return std::vector<float>(M * N, 0.0f);
        }
        auto [x_q, x_scale] = quantize_activation_per_token(x, M, K);
        std::vector<int8_t> w_q = unpack_awq_qweight(qweight, K, n_over_8);
        std::vector<int8_t> w_z = unpack_awq_qzeros(qzeros, groups, n_over_8);

        std::vector<float> out_fp32(M * N, 0.0f);
       
        for(int g = 0; g < groups; g++) {
            int k0 = g * group_size;
            // int k1 = (g + 1) * group_size;
            std::vector<int32_t> w_block(group_size * N);
            for(int k = 0; k < group_size; k++) {
                for(int n = 0; n < N; n++) {
                    w_block[k * N + n] = static_cast<int32_t>(w_q[(k0 + k) * N + n]) - static_cast<int32_t>(w_z[g * N + n]);
                }
            }

            std::vector<int32_t> x_block(M * group_size);
            for(int m = 0; m < M; m++) {
                for(int k = 0; k < group_size; k++) {
                    x_block[m * group_size + k] = static_cast<int32_t>(x_q[m * K + k0 + k]);
                }
            }
                //x_block = x_q[:, k0:k1].to(torch.int32) 
            std::vector<int32_t> acc(M * N);
            gemm_int32(x_block.data(), w_block.data(), acc.data(), M, group_size, N);

            for(int m = 0; m < M; m++) {
                for(int n = 0; n < N; n++) {
                    out_fp32[m * N + n] += static_cast<float>(acc[m * N + n]) * static_cast<float>(scales[g * N + n]);
                }
            }
        }
        for(int m = 0; m < M; m++) {
            for(int n = 0; n < N; n++) {
                out_fp32[m * N + n] *= x_scale[m];
            }
        }
        if(bias != nullptr) {
            if(bias.size() != N) {
                throw std::runtime_error("bias 形状应为: " + std::to_string(bias.size()) + " 当前： " + std::to_string(N));
            }
            for(int m = 0; m < M; m++) {
                for(int n = 0; n < N; n++) {
                    out_fp32[m * N + n] += static_cast<float>(bias[n]);
                }
            }
        }
        return out_fp32;
    }
}